Fix converting a reusable block with nested blocks into a static block

noisysocks opened the pull request:
Fixes #9278.

The `CONVERT_BLOCK_TO_STATIC` effect needs to handle the case where the block being converted has inner blocks.

#### Testing

1. Insert a Columns block.
2. Insert some other block into the Columns block.
3. Open the ellipsis (<kbd><samp>More Options</samp></kbd>) menu and choose <kbd><samp>Add to Reusable Blocks</samp></kbd>.
4. Finish naming and saving the Reusable Block.
5. Notice that the nested blocks are saved in the Reusable Block as expected.
6. Open the ellipsis menu again and choose <kbd><samp>Convert to Regular Block</samp></kbd>.
7. The nested blocks should not have disappeared.

noisysocks committed:
Fix converting a reusable block with nested blocks into a static block

The CONVERT_BLOCK_TO_STATIC effect needs to handle the case where the
block being converted has inner blocks.

youknowriad commented on packages/editor/src/store/effects/reusable-blocks.js:
```diff
@@ -237,7 +237,11 @@ export const convertBlockToStatic = ( action, store ) => {
+		newBlocks = [ createBlock(
+			referencedBlock.name,
+			referencedBlock.attributes,
+			referencedBlock.innerBlocks
```
I think we should clone the blocks before calling this:

 - Insert the reusable block twice
 - Convert one of them to regular
 - Edit the regular blocks (notice they change in both blocks)

noisysocks committed:
Clone referenced block when converting a reusable block to static

By using cloneBlock(), we correctly make a copy of the innerBlocks which
means that updates to one block don't affect another.

slimmilkduds commented:
Can you please reconsider include this in 4.2? I know I’m been nagging about this issue and I promise this is the last time, whatever you decide 😁

youknowriad reviewed (approved):
LGTM 👍

talldan commented:
@noisysocks - sorry - was a bit slow to review this. The only thing I spotted is that I can now nest a reusable block inside of itself, which causes an infinite recursion and my chrome tab ran out of memory! 

I know, I know, but I had to try it! 🤣 

Want me to create an issue about that?

youknowriad commented:
@talldan I don't think it's related to this PR but I saw an issue about it (created yesterday I think)

talldan commented:
Thanks, found it - https://github.com/WordPress/gutenberg/issues/11206
